src/main/java/com/direwolf20/buildinggadgets/common/building/placement/CapabilityBlockProvider.java

```diff
@@ -0,0 +1,54 @@
+package com.direwolf20.buildinggadgets.common.building.placement;
+
+import com.direwolf20.buildinggadgets.common.building.implementation.SingleTypeProvider;
+import net.minecraft.init.Blocks;
+import net.minecraft.nbt.NBTBase;
+import net.minecraft.nbt.NBTTagCompound;
+import net.minecraft.nbt.NBTUtil;
+import net.minecraft.util.EnumFacing;
+import net.minecraftforge.common.capabilities.Capability;
+import net.minecraftforge.common.capabilities.Capability.IStorage;
+import net.minecraftforge.common.capabilities.CapabilityInject;
+import net.minecraftforge.common.capabilities.CapabilityManager;
+
+import javax.annotation.Nonnull;
+
+public class CapabilityBlockProvider {
+
+    @CapabilityInject(IBlockProvider.class)
+    public static Capability<IBlockProvider> BLOCK_PROVIDER = null;
+
+    static IBlockProvider DEFAULT_AIR_PROVIDER = new SingleTypeProvider(Blocks.AIR.getDefaultState());
+
+    public CapabilityBlockProvider() {}
+
+    public static void register() {
+        CapabilityManager.INSTANCE.register(IBlockProvider.class, new IStorage<IBlockProvider>() {
+            @Nonnull
+            @Override
+            public NBTBase writeNBT(Capability<IBlockProvider> capability, IBlockProvider provider, EnumFacing facing) {
+                if (provider instanceof OriginWrapper) {
+                    provider = ((OriginWrapper) provider).getHandle();
```

Comment:
Notice that the Handle maybe an OriginWrapper again... So if you'd just allowOriginWrapper to correctly delegate trhough...
(If you make the CapabilityInterface extend INBTSerializable<NBTTagCompound> this can be achieved easily)

Reply:
Don't think I'll implement this.
If I would do this then block providers won't immutable.

Reply:
You're right... But there are ways to do that so that you can keep them immutable: just make a read and write Method, which always returns a Object of the same type - the underlying implementation can then decide to be immutable

Reply:
Note that the methods supplied by Capability `void readNBT(Capability<T>, T, EnumFacing, NBTBase)` returns *void*, which means the only way to make this work is by modifying the parameter (which we can't do right now either, just realized)

Reply:
It is `void readNBT`, which means the only way to obey the contract is to modify the parameter, which we can't do with that (and right now...)

Reply:
Duplicate